backend/src/utils/chat.ts

```diff
@@ -12,19 +12,26 @@ function pushMessageToHistory(
 ) {
 	// limit the length of the chat history
 	const maxChatHistoryLength = 1000;
-	const updatedChatHistory = [...chatHistory];
 
-	// remove the oldest message, not including system role message
-	// until the length of the chat history is less than maxChatHistoryLength
-	while (updatedChatHistory.length >= maxChatHistoryLength) {
-		if (updatedChatHistory[0].chatMessageType === CHAT_MESSAGE_TYPE.SYSTEM) {
-			updatedChatHistory.shift();
-		} else {
-			updatedChatHistory.splice(1, 1);
-		}
+	const existingSystemRole =
+		chatHistory[0]?.chatMessageType === CHAT_MESSAGE_TYPE.SYSTEM
+			? chatHistory[0]
+			: null;
+
+	const limitedHistoryStartIndex = Math.max(
+		chatHistory.length - maxChatHistoryLength + 1,
+		0
+	);
+	const updatedChatHistory = chatHistory.slice(
+		limitedHistoryStartIndex,
+		chatHistory.length
+	);
+
+	if (existingSystemRole) {
+		updatedChatHistory[0] = existingSystemRole;
```

Comment:
Can't help thinking this is a touch more complex, not less. Within the scope of this function it's ok to be "mutating" our array copy to pop off messages, because all the logic is self-contained.

If you are intent on removing all the extra messages in one operation, rather than one at a time in a while loop, then how about this?

```
const updatedChatHistory = [...chatHistory, newMessage];
const messagesToRemove = updatedChatHistory.length - maxMessageLength;
if (messagesToRemove < 1) return updatedChatHistory;

const spliceFrom = updatedChatHistory[0].chatMessageType === CHAT_MESSAGE_TYPE.SYSTEM ? 1 : 0;
updatedChatHistory.splice(spliceFrom, messagesToRemove);
return updatedChatHistory;
```

Reply:
tasty